pallets/block-rewards/src/lib.rs

```diff
@@ -0,0 +1,449 @@
+			origin: OriginFor<T>,
+			currency_id: T::CurrencyId,
+			account_id: T::AccountId,
+		) -> DispatchResult {
+			ensure_signed(origin)?;
+
+			T::Rewards::claim_reward((T::Domain::get(), currency_id), &account_id).map(|_| ())
+		}
+
+		/// Admin method to set the reward amount for a collator used for the next epochs.
+		/// Current epoch is not affected by this call.
+		#[pallet::weight(T::WeightInfo::set_distributed_reward())]
+		pub fn set_collator_reward(
+			origin: OriginFor<T>,
+			collator_reward_per_session: T::Balance,
+		) -> DispatchResult {
+			T::AdminOrigin::ensure_origin(origin)?;
+
+			NextEpochChanges::<T>::mutate(|changes| {
+				changes.collator_reward = Some(collator_reward_per_session);
+			});
+
+			Ok(())
+		}
+
+		/// Admin method to set the inflation amount for the Beneficiary used for the next epochs.
+		/// Current epoch is not affected by this call.
+		#[pallet::weight(T::WeightInfo::set_distributed_reward())]
+		pub fn set_beneficiary_reward(
+			origin: OriginFor<T>,
+			beneficiary_reward_per_session: T::Balance,
+		) -> DispatchResult {
+			T::AdminOrigin::ensure_origin(origin)?;
+
+			NextEpochChanges::<T>::mutate(|changes| {
+				changes.beneficiary_reward = Some(beneficiary_reward_per_session)
+			});
+
+			Ok(())
+		}
+	}
+}
+
+impl<T: Config> Pallet<T> {
+	/// Mint default amount of stake for target address and deposit stake.
+	/// Enables receiving rewards onwards.
+	fn do_init_collator(who: &T::AccountId) -> DispatchResult {
+		T::Currency::mint_into(
+			T::CollatorCurrencyId::get(),
+			who,
+			T::DefaultCollatorStake::get(),
+		)?;
+		T::Rewards::deposit_stake(
+			(T::Domain::get(), T::CollatorCurrencyId::get()),
+			who,
+			T::DefaultCollatorStake::get(),
+		)
+	}
+
+	/// Withdraw currently staked amount for target address and immediately burn it.
+	/// Disables receiving rewards onwards.
+	fn do_exit_collator(who: &T::AccountId) -> DispatchResult {
+		let amount =
+			T::Rewards::account_stake((T::Domain::get(), T::CollatorCurrencyId::get()), who);
+		T::Rewards::withdraw_stake(
+			(T::Domain::get(), T::CollatorCurrencyId::get()),
+			who,
+			amount,
+		)?;
+		T::Currency::burn_from(T::CollatorCurrencyId::get(), who, amount).map(|_| ())
+	}
+
+	/// Apply epoch changes and distribute rewards.
+	///
+	/// NOTE: Noop if any call fails.
+	fn do_advance_epoch() -> Weight {
+		let mut num_collators = 0;
+		let mut num_joining = 0;
+		let mut num_leaving = 0;
+
+		transactional::with_storage_layer(|| -> DispatchResult {
+			NextEpochChanges::<T>::try_mutate(|changes| -> DispatchResult {
+				ActiveEpochData::<T>::try_mutate(|epoch_data| {
+					num_joining = changes.collators.inc.len();
+					num_leaving = changes.collators.out.len();
+
+					// Apply collator set changes before rewarding
+					for leaving in changes.collators.out.iter() {
+						Self::do_exit_collator(leaving)?;
+					}
+					for joining in changes.collators.inc.iter() {
+						Self::do_init_collator(joining)?;
```

Comment:
In the current design, all collators receive the same share of the reward. This is due to the fact that everyone has the same artificial stake invoked by `do_init_collators`. After removing the ability to adjust a single collator's stake via the `AdminOrigin`, it's currently not possible for collators to receive different amounts of rewards per epoc. Please see [my open question in the spec](https://centrifuge.hackmd.io/9T9zAZaTRD6ZcHN-Gd_lzw#Open-question) for details.

Reply:
I agree that every collator should get the same reward per epoch, so for me we do not need to change that behavior. What I am not sure I understand, is that a new collator gets the old (default) collator reward and not the new updated one from storage. Since the collator reward is per collator and not per epoch for all collators.

Reply:
There is no _default_ or _old collator reward_, only the `EpochData.collator_reward` stored in the `ActiveEpochData ` map as of now which denotes the reward each collator should receive per epoch. Thus, in total `num_collators * EpochData.collator_reward` for the group. For one epoch, all current collators receive exactly that amount, independent of whether they are new or not.

If in `session(i)` a change to the collator reward is configured and thus stored in `NextEpochChanges` StorageMap, then this change is applied in `session(i+1)` but AFTER rewarding the collators, so it becomes effective in `session(i+2)`. The argument of @lemunozm was

> I think it makes sense to reward first and change groups after. The main idea behind this is that once the epoch starts, nothing should change for that epoch, so I, as an investor, can evaluate if it's ok to leave my money there during the next whole period. If we allow updating the group before, an admin could change it 3 days before the 1-month epoch ends. Knowing that, as an investor, 1 month ago, I could have decided to not invest in that.
So, yes, you need at least 1 completed epoch to change group properties.

Reply:
ah! the `do_init_collator` sets/mints only the virtual stake. I thought it was already rewarding. Makes sense now.